### PR TITLE
Add Finetuning Scripts

### PR DESCRIPTION
**Pull request by skryl**
# CSM-MLX Finetuning

This module provides functionality for finetuning CSM (Conversation Speech Model) on custom text-audio datasets. Finetuning allows you to adapt the model to specific speakers, domains, or styles.

## Prerequisites

You'll need the following additional dependencies:
```bash
pip install tqdm audiofile audresample
```

## Preparing Your Dataset

To finetune the model, you need to prepare a dataset of text and audio pairs. Each sample should have:
- Text transcript
- Audio file (24kHz WAV recommended)
- Speaker ID (optional, but highly recommended for voice cloning)

### Using the Dataset Creation Script

We provide a helper script to create a dataset from your audio files with matching text files:

```bash
python -m csm_mlx.finetune.create_dataset \
  --audio-dir /path/to/audio/files \
  --output-json dataset.json
```

This script will automatically look for matching `.normalized.txt` or `.original.txt` files with the same basename as your audio files.

### Manual Dataset Creation

You can also create your dataset manually as a JSON file:
```json
[
  {
    "text": "This is an example.",
    "audio_path": "/path/to/audio1.wav",
    "speaker_id": 0
  },
  {
    "text": "Another example with a different speaker.",
    "audio_path": "/path/to/audio2.wav",
    "speaker_id": 1
  }
]
```

### LoRA Finetuning (Recommended)

LoRA (Low-Rank Adaptation) is a parameter-efficient fine-tuning method that significantly reduces memory requirements and training time while maintaining quality. Instead of updating all weights, LoRA injects trainable rank decomposition matrices into specific layers.

```bash
python -m csm_mlx.finetune.finetune_lora \
  --data-path dataset.json \
  --output-dir ./lora_finetune_output \
  --batch-size 4 \
  --epochs 5 \
  --learning-rate 5e-4 \
  --lora-rank 8 \
  --optimizer adamw \
  --target-modules attn codebook0_head projection \
  --train-embeddings
```

#### LoRA Parameters

- `--lora-rank`: Rank of the low-rank matrices (typically 4-16; higher rank = more capacity)
- `--lora-alpha`: Scaling factor (default is 16.0, often set to 2× rank)
- `--target-modules`: Which modules to apply LoRA to (defaults to attention, codebook head, and projection)
- `--train-embeddings`: Trains the embedding layers directly (not via LoRA adapters)

> **Important**: Do not include embedding layers in `--target-modules` when using `--train-embeddings`. These are mutually exclusive approaches for embedding layers. The `--train-embeddings` flag trains the full embedding matrices directly, while including them in `target_modules` would apply LoRA to them instead.

#### Memory Usage Considerations

There are three approaches for training embeddings, with different memory requirements:

1. **LoRA for everything (lowest memory)**: Don't use `--train-embeddings`
   ```
   --target-modules attn codebook0_head projection
   ```

2. **LoRA + direct embedding training (medium memory)**: Use `--train-embeddings` and don't include embeddings in `--target-modules`
   ```
   --target-modules attn codebook0_head projection --train-embeddings
   ```

3. **Full finetuning (highest memory)**: Use `csm_mlx.finetune.finetune` instead of LoRA

For best results with reasonable memory usage, option 2 is recommended.

### Full Finetuning

Full finetuning updates all model parameters. This gives maximum flexibility but requires more data and computational resources.

```bash
python -m csm_mlx.finetune.finetune \
  --data-path dataset.json \
  --output-dir ./finetune_output \
  --batch-size 4 \
  --epochs 5 \
  --learning-rate 1e-5
```

### CODE REVIEW

**Comment by senstella** [comment]
Hello, this is awesome! Thank you so much for sending this PR. It looks very promising and fascinating! I will get to testing and reviewing shortly. Really appreciate it!

File: csm_mlx/finetune/trainer.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        c0_logits = self.model.codebook0_head(backbone_hidden)
+        c0_loss = nn.losses.cross_entropy(
+            c0_logits.reshape(-1, c0_logits.shape[-1]),
+            audio_targets[:, :, 0].reshape(-1),
+            reduction="none"
+        )
+        c0_loss = (c0_loss * audio_masks[:, :, 0].reshape(-1)).sum() / audio_masks[:, :, 0].sum()
+        
+        # Other codebooks (through decoder)
+        decoder_inputs = self.model.projection(backbone_hidden)
```
Comment (senstella) [inline csm_mlx/finetune/trainer.py]: I think `self.model.projection` input should be a combination of `backbone_hidden` and list of `ci_logit`s, such as in inference:
```py
        c0_logprobs = c0_logits - mx.logsumexp(c0_logits, keepdims=True)
        c0_sample = mx.expand_dims(sampler(c0_logprobs), axis=-1)
        c0_embeds = model.embed_audio(0, c0_sample)

        decoder_inputs = mx.concat(
            [mx.expand_dims(backbone_last_hidden, axis=1), c0_embeds], axis=1
        )
```
It might be more simple to fine-tune backbone and decoder seperately - tune backbone for voice memorization and tune decoder to enhance quality of it.

File: csm_mlx/finetune/lora.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+        # Method 2: Try direct MLX state dict loading
+        lambda: (
+            print("Trying MLX state dict loading..."),
+            mx.load(checkpoint_path)
+        ),
+        
+        # Method 3: Try safetensors loading
+        lambda: (
+            print("Trying safetensors loading..."),
+            load_safetensors(checkpoint_path)
```
Comment (senstella) [inline csm_mlx/finetune/lora.py]: MLX supports safetensors natively with both [`nn.load_weights`](https://ml-explore.github.io/mlx/build/html/python/nn/_autosummary/mlx.nn.Module.load_weights.html) and [`mx.load`](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.load.html). So, I think this `Method 3` can be removed safely.

File: csm_mlx/finetune/lora.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    if not lora_weights:
+        print("Warning: No LoRA parameters found in the checkpoint")
+    
+    # Load embedding weights if requested
+    if load_embeddings and embedding_weights:
+        all_params = model.parameters()
+        for key, value in embedding_weights.items():
+            if key in all_params:
+                # Convert to MLX array if needed
+                if not isinstance(value, mx.array):
```
Comment (senstella) [inline csm_mlx/finetune/lora.py]: If we remove `Method 3` above, I think this line also can be safely removed - since both methods return MLX arrays.

File: csm_mlx/finetune/lora.py
Comment (senstella) [inline csm_mlx/finetune/lora.py:1]: Generally, I think it would be simpler to maintain and extend if we re-use [`mlx-lm`'s LoRA](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/tuner/lora.py).

File: csm_mlx/finetune/trainer.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        decoder_hidden = self.model.decoder(decoder_inputs)
+        
+        other_codebooks_loss = 0.0
+        for i in range(1, self.model.n_audio_codebooks):
+            ci_logits = mx.matmul(decoder_hidden, self.model.audio_head[i-1])
```
Comment (senstella) [inline csm_mlx/finetune/trainer.py]: I think we need to use `[backbone_last_hidden, c0_embeds, c1_embeds, ...]` as the input x for `decoder(projection(x))` to obtain the activations. For the activations at indices `[2:]`, we should perform a matrix multiplication with the audio_head for each index i, and then calculate the losses for these.

**Review by senstella — CHANGES_REQUESTED** [review]

**Comment by senstella** [comment]
I've briefly reviewed the code and it looks quite promising! Most of it seems pretty solid, but I noticed a few issues specifically related to the loss calculations of decoder. Could you please take a look at the comments I've left regarding those? Thank you!

**Comment by senstella** [comment]
Hello, @skryl!

Since it's really nice feature and I'm excited to get it merged, I'm thinking I'll make a small commits to fix the issues I've mentioned and restructure a bit to fit better in the current structure of the repo. This way, I think we can get this PR merged sooner.

I'm planning to start making these changes in a few hours. If you'd prefer to handle them yourself instead, just let me know and I'll step back!

Again, I really appreciate the contribution to the project! Thanks!

**Comment by skryl** [comment]
@senstella sounds good, go for it! I probably won't have time to look at it until tomorrow anyways. Also, if mlx already has a lora library then totally makes sense to use it, I just couldn't find one when I looked.

File: csm_mlx/finetune/trainer.py (hunk 4 above)
Comment (skryl) [reply csm_mlx/finetune/trainer.py]: Makes sense, I wasn't sure what loss was used for training so just kinda guessed here.

**Comment by senstella** [comment]
@skryl How do you think about instead of using TRL-styled Trainer class inheritance in [`trainer.py`](https://github.com/senstella/csm-mlx/blob/b273612c51c2f5dc90b4736f484fd72202b1ff65/csm_mlx/finetune/trainer.py), using [`mlx-lm` style single function trainer](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/tuner/trainer.py)? I think it would be better practice to use mlx style trainer, since this is an mlx repository. Also, `mlx-lm` style trainer includes `loss: callable = default_loss,` as a parameter rather than a class method. Personally, I think it's easier for people to experiment with custom loss functions than overriding from a class - such as DPO, KTO and more.

About training config, instead of getting args separately in `__init__`, I think it would be more extendable in the future - such as passing args between functions/classes - to use a `@dataclass` approach like TRL and `mlx-lm` do, regardless of whether it's a class-style trainer or single-function-style trainer!

Please let me know what you think!

**Comment by senstella** [comment]
Also, I think it's cleaner to remove datasets and example directories from the repository and add something like `FINETUNING.md` or add a `docs/` directory to explain how to do things - to keep things clean for people who use this library only for inference purposes.

Regarding the training dataset format, I think it would be better practice to use `List[List[Segment]]` to represent a dataset since it's a conversational model rather than a normal TTS model. `AudioTextSample` feels a little redundant to me. And we might add some options such as `mask_prompt` to prevent the model from learning speakers we don't want it to!

Additionally, it would be wonderful to add [`datasets`](https://huggingface.co/docs/datasets/index) support to the `CSMDataset` like how [`mlx-lm`'s tuner does](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/tuner/datasets.py).

However, since the changes I've suggested introduce significant architectural changes to the PR, I wonder what your thoughts are regarding these changes!

**Comment by senstella** [comment]
Here's a brief update! I've updated the dataset format to `List[List[AudioTextSample]]` since this is the correct prompting template for the CSM model, and added a loss mask to ignore specific `speaker_id`s. I'll move the CLI script to the `csm_mlx/cli` directory and use `typer` to match the current project structure. Let me know if you prefer a different approach!

**Comment by skryl** [comment]
Awesome, can't wait to try it out!

**Comment by swolf0** [comment]
Howdy!

Thanks to both of you for working on this! I've been having a lot of fun messing with it in my spare time.

I've been trying to tune the attention, codebook head, and projection modules on an impromptu data set. The tune successfully runs, taking up to ~91GB of RAM on my system! However, loading the finetune has been a bit of an issue. I've tried to hack the `example_finetuned_lora.py` function into working using `linear_to_lora_layers` from finetune/utils, but I'm not sure that's correct.

Yesterday, it seemed like the tune had no effect, but today I'm getting garbled audio.

I tuned with:
```
python -m csm_mlx finetune lora \
  --data-path ./datasets/dataset_updated.json \
  --output-dir ./lora_finetune_output \
  --target-modules attn codebook0_head projection
```

I'm happy to provide more info (though I'm not sure how useful it will be) or run any test sets.
I plan to edit the READMEs with the new CLI commands tomorrow should I get some time!

**Comment by senstella** [comment]
Hello, thank you for testing it out!

I think the correct way to load the trained LoRA is:
```python3
# Assume you already initialized CSM and loaded the weights
from csm_mlx.finetune.utils import linear_to_lora_layers
linear_to_lora_layers(
    csm,
    {
        "rank": 8,
        "scale": 2.0,
        "dropout": 0.0,  # TODO: Add dropout rate
        "keys": {"self_attn.q_proj", "self_attn.v_proj", "codebook0_head", "projection"},
        # Not using `attn` since there'd be some API changes with more projections (like self_attn.k_proj) & MLP assigned for `attn`!
    },
)
csm.load_weights("./lora_finetune_output/final_model.safetensors")

# Inference?
```
for now. I'm thinking of making a small change to the saving structure, like adding an option to 'fuse' the LoRA layer into the model when the training run is over.

It's interesting that it used 91GB of RAM for just LoRA finetuning. Perhaps that's because the decoder part is training fully on the whole sequence for each batch. The Sesame blog mentions this, using compute amortization (training only on 1/16 of the sequence) to resolve this issue. I'll apply that to the current codebase after a bit of code restructuring on the Trainer.

Also, getting garbled audio with LoRA is pretty weird. It could be a problem with the current `attn` LoRA setup, which only trains on `self_attn.q_proj` and `self_attn.v_proj` for `attn` paramater, or perhaps some exposure biases in loss function - might need arguments like dropout.
I'll change the default to include all projection layers in attention and the MLP. It would be really helpful for me to figure out what's wrong with the code if you could share more details, and perhaps the loss logs too!

**Comment by swolf0** [comment]
I'm still working on building out my synthetic data generator, but I ran it with an initial batch of 25 samples at 10s each:

Here's the run (reaching 91.72 GB RAM at commit 61c28ee1):
https://gist.github.com/swolf-personal/12b53e0c9476572efb990d85cf66eff9

And here's what I was using to try and load the layers (which failed):
https://gist.github.com/swolf-personal/6dc528eee99a8dedc3b081c246707e43

Please let me know if I am missing any data you need for debugging or want me to try anything specific!

**Comment by senstella** [comment]
That's interesting, thank you for providing the detailed logs! Please note that I introduced more LoRA layers(such as proj_k) as the default for `attn` selector in 61c28ee.

Thus, correct code to load the freshly trained model in 61c28ee should be:

```python3
from csm_mlx.finetune.utils import linear_to_lora_layers
linear_to_lora_layers(
    csm,
    {
        "rank": 8,
        "scale": 2.0,
        "dropout": 0.0,  # TODO: Add dropout rate
        "keys": {"attn", "codebook0_head", "projection"},
    },
)
csm.load_weights("./lora_finetune_output/final_model.safetensors")

# Inference?
```

Please let me know if this correctly loads the LoRA model for you!

**Comment by swolf0** [comment]
Oops 😅 seems obvious in hindsight, thanks! 

Everything is loading correctly now, and I'm really surprised at the quality. It cloned Kokoro's heart voice very convincingly and most importantly seems to be stable when quantized with `nn.quantize(csm)`. I can get realtime on my system, but only at 8 bit or less quantization. One quirk is that the new voice can't seem to say "fine", but it wasn't in my dataset at all, so that's probably reasonable.

Thanks again for your work!

**Comment by senstella** [comment]
Really glad to hear that it worked well!

Regarding the issue that the model can't say the word "fine", I suspect that it somewhat overfitted to your dataset. Perhaps hyper-parameter tuning would help a bit! I'll add some offline RL-based tuning methods like DPO and KTO later once this PR is complete. Perhaps it will generalize better than SFT!

Thank you for experimenting with it!

**Comment by senstella** [comment]
@skryl I've made significant modifications to Trainer, moving state to `TrainerState` and `History`, using `TrainArgs` instead of getting arguments in `__init__()`, and finally using `CheckpointManager` to manage checkpoint saving and loading for better training resumption! I'd love to hear your thoughts on these changes and any suggestions you might have for further improvements!

I think now we just need to add some documentation and a few more constructors for CSMDataset, and then should be all set to merge!

**Comment by senstella** [comment]
I wanted to let you know that I've finally completed the refining process and merged the PR into the master branch. I really appreciate all the work you put into this PR! I've also added some new documentation to explain the fine-tuning process! Again, thank you for all your contributions with this long PR! Please let me know if you find any bugs.

**Comment by sjlynch** [comment]
Unfortunately I'm on Windows, would this work in docker?

**Comment by senstella** [comment]
> Unfortunately I'm on Windows, would this work in docker?

MLX added cuda support on Linux, so docker with Linux might work with it. Though not stable as Mac version does. However, [HF added a trainer for CSM](https://huggingface.co/docs/transformers/model_doc/csm#training), so if you don't need those experimental offline-RL features, you might want to directly use it!
